python/cuml/dask/common/input_utils.py

```diff
@@ -0,0 +1,314 @@
+
+        self.total_rows = 0
+
+        self.parts_to_sizes = dict()
+
+        # func = _get_rows_from_colocated_tuple if self.multiple is True else \
+        #     _get_rows_from_single_obj
+
+        key = uuid1()
+
+        if self.multiple:
+            for w, futures in self.gpu_futures.items():
+                parts = [(self.client.submit(
+                    _get_rows_from_colocated_tuple,
+                    future,
+                    workers=[w],
+                    key="%s-%s" % (key, idx)).result())
+                    for idx, future in enumerate(futures)]
+
+                self.parts_to_sizes[self.worker_info[w]["r"]] = parts
+                for p in parts:
+                    self.total_rows = self.total_rows + p
+
+        else:
+                self.parts_to_sizes = [(ranks[wf[0]], self.client.submit(
+                                       _get_rows_from_single_obj,
+                                       wf[1],
+                                       workers=[wf[0]],
+                                       key="%s-%s" % (key, idx)).result())
+                                       for idx,
+                                       wf in enumerate(self.gpu_futures)]
+
+                self.total_rows = reduce(lambda a, b:
+                                         a + b, map(lambda x: x[1],
+                                                    self.parts_to_sizes))
+
+
+@with_cupy_rmm
+def concatenate(objs, axis=0):
+    if isinstance(objs[0], DataFrame):
+        if len(objs) == 1:
+            return objs[0]
+        else:
+            return cudf.concat(objs)
+
+    elif isinstance(objs[0], cp.ndarray):
+        return cp.concatenate(objs, axis=axis)
+
+
+def to_output(futures, type, client=None, verbose=False):
+    if type == 'cupy':
+        return to_dask_cupy(futures, client=client, verbose=verbose)
+    else:
+        return to_dask_cudf(futures, client=client, verbose=verbose)
+
+
+def _get_meta(df):
+    """
+    Return the metadata from a single dataframe
+    :param df: cudf.dataframe
+    :return: Row data from the first row of the dataframe
+    """
+    ret = df[0].iloc[:0]
+    return ret
+
+
+def _to_dask_cudf(futures, client=None, verbose=False):
+    """
+    Convert a list of futures containing cudf Dataframes into a Dask.Dataframe
+    :param futures: list[cudf.Dataframe] list of futures containing dataframes
+    :param client: dask.distributed.Client Optional client to use
+    :return: dask.Dataframe a dask.Dataframe
+    """
+    c = default_client() if client is None else client
+    # Convert a list of futures containing dfs back into a dask_cudf
+    dfs = [d for d in futures if d.type != type(None)]  # NOQA
+    if verbose:
+        print("to_dask_cudf dfs=%s" % str(dfs))
+    meta = c.submit(_get_meta, dfs[0]).result()
+    return dd.from_delayed(dfs, meta=meta)
+
+
+""" Internal methods, API subject to change """
+
+
+@gen.coroutine
+def _extract_partitions(dask_obj, client=None):
+
```

Comment:
One thing to think about with this whole API is how can we better communicate in code which functions return futures vs. forcing compute vs. persisting stuff. I often find it hard to look at code with dask wrappers and reason about what exactly I should expect to get back. The more we can lean into leaving things delayed until we actually need it, the happier we'll be (or at least the happier Corey will be!), I assume...

Reply:
What if a function like this were called `_compute_partitions` or something similar to make it obvious that it's computing everything?